vm/systemSmartContracts/stakingWaitingList.go

```diff
@@ -0,0 +1,1170 @@
+const waitingElementPrefix = "w_"
+
+type waitingListReturnData struct {
+	blsKeys         [][]byte
+	stakedDataList  []*StakedDataV2_0
+	lastKey         []byte
+	afterLastJailed bool
+}
+
+func (s *stakingSC) processStake(blsKey []byte, registrationData *StakedDataV2_0, addFirst bool) error {
+	if registrationData.Staked {
+		return nil
+	}
+
+	registrationData.RegisterNonce = s.eei.BlockChainHook().CurrentNonce()
+	if !s.canStake() {
+		s.eei.AddReturnMessage(fmt.Sprintf("staking is full key put into waiting list %s", hex.EncodeToString(blsKey)))
+		err := s.addToWaitingList(blsKey, addFirst)
+		if err != nil {
+			s.eei.AddReturnMessage("error while adding to waiting")
+			return err
+		}
+		registrationData.Waiting = true
+		s.eei.Finish([]byte{waiting})
+		return nil
+	}
+
+	if !s.flagStakingV4.IsSet() {
+		err := s.removeFromWaitingList(blsKey)
+		if err != nil {
+			s.eei.AddReturnMessage("error while removing from waiting")
+			return err
+		}
+	}
+	s.addToStakedNodes(1)
+	s.activeStakingFor(registrationData)
+
+	return nil
+}
+
+func (s *stakingSC) unStakeAtEndOfEpoch(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, s.endOfEpochAccessAddr) {
+		// backward compatibility - no need for return message
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) != 1 {
+		s.eei.AddReturnMessage("not enough arguments, needed the BLS key")
+		return vmcommon.UserError
+	}
+
+	registrationData, err := s.getOrCreateRegisteredData(args.Arguments[0])
+	if err != nil {
+		s.eei.AddReturnMessage("cannot get or create registered data: error " + err.Error())
+		return vmcommon.UserError
+	}
+	if len(registrationData.RewardAddress) == 0 {
+		s.eei.AddReturnMessage("cannot unStake a key that is not registered")
+		return vmcommon.UserError
+	}
+	if registrationData.Jailed && !registrationData.Staked {
+		s.eei.AddReturnMessage("already unStaked at switchJailedToWaiting")
+		return vmcommon.Ok
+	}
+
+	if !registrationData.Staked && !registrationData.Waiting {
+		log.Debug("stakingSC.unStakeAtEndOfEpoch: cannot unStake node which was already unStaked", "blsKey", hex.EncodeToString(args.Arguments[0]))
+		return vmcommon.Ok
+	}
+
+	if registrationData.Staked {
+		s.removeFromStakedNodes()
+	}
+
+	if registrationData.Waiting {
+		err = s.removeFromWaitingList(args.Arguments[0])
+		if err != nil {
+			s.eei.AddReturnMessage(err.Error())
+			return vmcommon.UserError
+		}
+	}
+
+	registrationData.Staked = false
+	registrationData.UnStakedEpoch = s.eei.BlockChainHook().CurrentEpoch()
+	registrationData.UnStakedNonce = s.eei.BlockChainHook().CurrentNonce()
+	registrationData.Waiting = false
+
+	err = s.saveStakingData(args.Arguments[0], registrationData)
+	if err != nil {
+		s.eei.AddReturnMessage("cannot save staking data: error " + err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (s *stakingSC) unStake(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```

Comment:
unStake should be in the base file - not here.

Reply:
maybe make a refactor - unStakeV1 and unStakeV2 -> and call them depending on flag.

Reply:
Good suggestion with unStakeV1 & unStakeV2. Applied it !